keps/sig-node/4817-resource-claim-device-status/README.md

```diff
@@ -0,0 +1,516 @@
+
+This extension also lays the foundation for a potential future standardization of specific type device data, such as, for example, networking devices information.
+
+## Motivation
+
+As of now, when a device is configured in a pod/container, the state and characteristics of the device set during the configuration stage are invisible. The lack of this information is then a challenge for the users and developers to diagnose issues, verify configurations, and integrate the allocated resources into higher-level services. Reporting this information is then crucial in environments where device specific configurations are necessary.
+
+For certain types of devices, such as network interfaces, knowing the detailed characteristics of what has been allocated is particularly useful and crucial. For example, reporting the interface Name, MAC address and IP addresses of network interfaces in the status of a ResourceClaim can significantly help in configuring and managing network services, as well as in debugging network related issues. By including such device specific information, this proposal addresses existing gaps in visibility and facilitates better integration and management of resources.
+
+### Goals
+
+* Allow arbitrary, driver-specific information to be reported from the DRA drivers for each allocated device in a ResourceClaim.
+* Establish a foundation for potential future standardization (e.g. Network Devices status).
+* Enable 3rd party implementation of new functionalities based on the Device Status (e.g. Secondary Network Services if the IPs of a network device are reported).
+
+### Non-Goals
+
+* Implement new functionalities based on the Device Status (e.g. Kubernetes EndpointSlice Controller supporting IPs reported by networking DRA drivers).
+* Modifying the Resource Claim workflow to include the device status report.
+
+## Proposal
+
+### API - ResourceClaim.Status
+
+The API changes define a new `DeviceStatuses` field in the existing `ResourceClaimStatus` struct. `DeviceStatuses` is a slice of a new struct `AllocatedDeviceStatus` which holds device specific information. 
+
+A device, identified by `<driver name>/<pool name>/<device name>` can be represented only once in the `DeviceStatuses` slice and will also mention which request caused the device to be allocated. The state and characteristics of the device are reported in the `Conditions`, representing the operational state of the device and in the `DeviceInfo`, an arbitrary data slice representing device specific characteristics. Additionally, for networking devices, a field `NetworkDeviceInfo` can be used to report the IPs, the MAC address and the interface name.
```

Comment:
This is one approach, where there is `DeviceInfo` and `NetworkDeviceInfo`. I am not sure that will pass API review muster, but maybe. An alternative is what is described in Option 5. Instead of 3 fields, there would be two fields: `Conditions` and `DeviceInfo`. The `DeviceInfo` field is a one-of API type. So, something like:

```go
type AllocatedDeviceStatus struct {
  ...
  DeviceInfo *DeviceInfo
}

type DeviceInfo struct {
    // +optional
	// +oneOf=DeviceInfoType
    Network *DeviceInfoNetwork

   // +optional
   // +oneOf=DeviceInfoType
   General *DeviceInfoGeneral
}

type DeviceInfoGeneral struct {
  Opaque RawExtension
}

type DeviceInfoNetwork struct {
   InterfaceName string
   IPAddresses []IPAddress
   MACAddress string

  Opaque RawExtension
}
```

@thockin WDYT? *Every* one-of would have the `Opaque RawExtension` plus one-of specific fields. Should that therefore just bubble up one level, so that we have a RawExtension, and then an optional device-type based one-of? Or, what about compound devices that may be several types. Should we just make the `Opaque` and `Network` fields separate, optional fields that may or may not be populated for any given device? I am not sure the one-of construction actually helps. (In other words, go with what is already here in the KEP).

Reply:
I see, I interpreted it like any driver can use the opaque field (`DeviceInfo`) + their own type. So the `oneOf` would be between specific type (`NetworkDeviceInfo`, `GPUDeviceInfo`...).

Reply:
Yeah, but the way you did it is less prescriptive - it allows a device to serve multiple functions. So I rather like that.

Reply:
To be a little more verbose about my last comment:

1) we don't know what devices will come out in the future. So, whose to say they won't stick an optical connection on a GPU...or maybe they have already? So, maybe the GPU and NIC will be a single device, not two separate devices

2) As VM architects, there may be specific devices that "go together as one". The GPU/NIC combos is an example. Right now, we are using MatchAttributes with a pcieRoot attribute to align the internal topology. That's a lot of work for the user. Instead, I am looking at building a "compound device driver" on top of the partitionable device model. This would allow a single "device" owned by that driver to be allocated, but in reality that driver would proxy calls for that device to multiple other drivers. So, the VM architect could "pair" specific devices, such that you cannot get one without the other. Those devices would have separate functions, represented by the separate underlying drivers. I would want to be able to stick information about each underlying device in the right "device info field" for that function.

Thinking about it, this is still limited of course, to one device per function. Another option would be to have the one-of, but allow multiple entries per driver/pool/device tuple.